### PR TITLE
Upgrade book to v0.15

### PR DESCRIPTION
**Pull request by eikooc**
Hello @JordanMartinez. 

We chatted on the PureScript Discourse about me opening a PR where we could continue the discussion.
I've gotten the book updated a bit and changed some of the upstream versions that were manually set to forks that don't look maintained but the main repositories does.

Right now the build fails on a missing package in the package set `optparse`, but I'll continue I just wanted to open the PR as discussed.

Looking forward to working some more with this.

### CODE REVIEW

**Comment by JordanMartinez** [comment]
Thanks for doing this!

File: exercises/chapter12/src/Example/LSystem.purs
Comment (ptrfrncsmrph) [inline exercises/chapter12/src/Example/LSystem.purs:10]: I know this probably keeps the diff minimal, but I think it's more idiomatic to have the namespace match the module name, like
```suggestion
import Data.Number as Number
```

Comment (eikooc) [reply exercises/chapter12/src/Example/LSystem.purs:10]: @JordanMartinez and @ptrfrncsmrph thanks for the reviews.
When is it idiomatic to use the namespace and when is it idiomatic to remove it entirely? I'd think it would be if only a few functions from the library are used. I was wondering if I should just drop the namespace and import `(cos, sin, tau)`?

Comment (JordanMartinez) [reply exercises/chapter12/src/Example/LSystem.purs:10]: I don't think there is an idiomatic usage of a namespace or not. Often times, the namespace is useful when you want to clarify _which_ function you're using. For example, I could do `length`, but it's not obvious _which_ length that is. If I use `Array.length` or `String.length`, it's pretty obvious.

I think keeping the `Number` namespace here is desirable because there are other versions of `cos` (e.g. `Data.Decimal (cos)`). It's also probably a smaller diff if just the namespace is changed.

File: exercises/chapter2/spago.dhall
Comment (thomashoneyman) [inline exercises/chapter2/spago.dhall:14]: This package is always included by Spago now, so it can be removed from the dependency lists of these various packages. Nothing needs to list it.

Comment (eikooc) [reply exercises/chapter2/spago.dhall:14]: Alright, I'll remove it then 👍

**Comment by kush87** [comment]
is it possible to merge this pull request now? Trying to learn the purescript from the book but the examples don't work with the compiler.

Thanks

**Review by JordanMartinez — CHANGES_REQUESTED** [review]
Looks like all of the package overrides/additions in the `chapter10/packages.dhall` file can be removed as they are part of the package set. I think Miles did those changes because he was using forks at the time.

**Comment by JordanMartinez** [comment]
> is it possible to merge this pull request now? Trying to learn the purescript from the book but the examples don't work with the compiler. It would be of great help as i am trying to introduce purescript into our organization.
> 
> Thanks

@kush87 This PR builds, so if you want to use an early version of the book, I'd use this PR's branch. The changes that still need to be done are minor (see above).

File: exercises/chapter3/packages.dhall
Comment (JordanMartinez) [reply exercises/chapter3/packages.dhall:18]: We don't need to override that. We just need to update the package set to a newer version:
```console
$ cat packages.dhall 
let upstream =
      https://github.com/purescript/package-sets/releases/download/psc-0.15.2-20220613/packages.dhall
        sha256:99f976d547980055179de2245e428f00212e36acd55d74144eab8ad8bf8570d8

in  upstream
$ spago ls packages | grep 'test'
test-unit                    v17.0.0   Remote "https://github.com/bodil/purescript-test-unit.git"
```
And to clarify, the package set is immutable. So, as long as the package set version isn't updated, that's the same thing as pinning the `test-unit` version to `v17.0.0`.

File: exercises/chapter2/spago.dhall
Comment (eikooc) [reply exercises/chapter2/spago.dhall:14]: @thomashoneyman it looks like it is still has to be a dependency as of `psc-0.15.2-20220613` since `spago install` fails with:
```
[error] Some of your project files import modules from packages that are not in the direct dependencies of your project.
To fix this error add the following packages to the list of dependencies in your config:
- numbers
You may add these dependencies by running the following command:
spago install numbers
```

**Comment by eikooc** [comment]
@JordanMartinez I've started removing all the additions and overrides but `react-basic-hooks` need to be in the additions until a release is made that updates the package to v8.0.0.

I have also removed the `psci-support` package as you mentioned.

There are also a lot of packages installed that are not used are they kept so someone solving the exercises can easily import what they need? Or can I just delete them?
As an example:
```
None of your project files import modules from some projects that are in the direct dependencies of your project.
These dependencies are unused. To fix this warning, remove the following packages from the list of dependencies in your config:
- functions
```

**Comment by JordanMartinez** [comment]
> @JordanMartinez I've started removing all the additions and overrides but `react-basic-hooks` need to be in the additions until a release is made that updates the package to v8.0.0.

Then we'll add it to the package set. See https://github.com/purescript/package-sets/pull/1173

> There are also a lot of packages installed that are not used. Are they kept so someone solving the exercises can easily import what they need? Or can I just delete them?

Let's assume that they are kept so someone can import what they need. Perhaps the book should mention this somewhere so people learn to ignore those warnings while they learn but know to drop unneeded packages when they're building real code?

File: exercises/chapter2/spago.dhall
Comment (thomashoneyman) [reply exercises/chapter2/spago.dhall]: Sorry, this diff is a little confusing. I commented on the line with “psci-support” and meant to refer to that package.

**Comment by JordanMartinez** [comment]
Once the next package set is released (tomorrow?), `react-basic-hooks` will be included in the package set and we can stop adding it here. Then, this PR should be good to merge.

File: exercises/chapter3/packages.dhall
Comment (milesfrain) [reply exercises/chapter3/packages.dhall:18]: I'm overriding the `test-unit` repo while waiting for this PR to be merged: https://github.com/bodil/purescript-test-unit/pull/47

**Comment by eikooc** [comment]
> Let's assume that they are kept so someone can import what they need. Perhaps the book should mention this somewhere so people learn to ignore those warnings while they learn but know to drop unneeded packages when they're building real code?

@JordanMartinez alright, so where should we take discussions around changes like that to the book? It would probably be best to at least be worked on in a separate pull request.

File: exercises/chapter3/packages.dhall
Comment (JordanMartinez) [reply exercises/chapter3/packages.dhall:18]: @milesfrain :wave: Hey! How's it been?

Looking at that PR, it's a purely cosmetic change (though it's one I agree with). It's not a blocker for this PR, but it would be a nice change to add to `test-unit`.

**Comment by JordanMartinez** [comment]
> > Let's assume that they are kept so someone can import what they need. Perhaps the book should mention this somewhere so people learn to ignore those warnings while they learn but know to drop unneeded packages when they're building real code?
> 
> @JordanMartinez alright, so where should we take discussions around changes like that to the book? It would probably be best to at least be worked on in a separate pull request.

Agreed. That could be done in a follow-up PR.

File: exercises/chapter2/spago.dhall
Comment (eikooc) [reply exercises/chapter2/spago.dhall]: Okay, whoops. Definitely thought you meant `numbers` 😅

File: exercises/chapter10/test/Examples.js
Comment (JordanMartinez) [inline exercises/chapter10/test/Examples.js:1]: Oo... I should have mentioned this earlier, but we don't need `"use strict";` anymore in ES modules because it's strict by default. That can be fixed in a separate PR thought so that we unblock this one.

File: exercises/chapter10/test/Examples.js
Comment (JordanMartinez) [inline exercises/chapter10/test/Examples.js:29]: There's an extra blank line here. That can be fixed in a future PR.

File: exercises/chapter12/spago.dhall
Comment (JordanMartinez) [inline exercises/chapter12/spago.dhall:13]: Was this removal intentional?

**Review by JordanMartinez — APPROVED** [review]

**Comment by JordanMartinez** [comment]
Thanks for all your help!

**Comment by JordanMartinez** [comment]
@thomashoneyman If you don't see anything else, mind if I merge this? Even if the comment above about the `list` dependency being removed?

**Comment by JordanMartinez** [comment]
Yup! That was my thought process as well.

**Comment by eikooc** [comment]
Thanks @thomashoneyman and @JordanMartinez for all the reviews. It's been a pleasure to work on!

File: exercises/chapter12/spago.dhall
Comment (eikooc) [reply exercises/chapter12/spago.dhall:13]: Non-intentional, I must've forgotten to add it back in. It was one of those packages that are not used, but still imported potentially for the reader's convenience when solving exercises.

**Pull request merged**
